Fix Twitter embeds

irystocratanon opened the pull request:
Twitframe has been broken for a long time now thanks to elon and Twitter's own embeds are also somewhat unreliable and don't work for NSFW tweets, etc.

Replace it with FxTwitter

Currently they don't have an iframe embed but they do have a simple JSON API that can be used instead.

https://github.com/FixTweet/FxTwitter/issues/774

irystocratanon committed:
Fix Twitter embeds

Twitframe has been broken for a long time now thanks to elon and
Twitter's own embeds are also somewhat unreliable and don't work for
NSFW tweets, etc.

Replace it with FxTwitter

Currently they don't have an iframe embed but they do have a simple JSON
API that can be used instead.

https://github.com/FixTweet/FxTwitter/issues/774

TuxedoTako commented on src/Linkification/Embedding.js:
```diff
@@ -592,30 +592,51 @@ var Embedding = {
-            $.off(window, 'message', onMessage);
-            return (cont || el).style.height = `${+$.minmax(e.data.height, 250, 0.8 * doc.clientHeight)}px`;
+          // Can't use fetch: I couldn't figure out how to get async to work here
+          // synchronous XHR is not necessarily the best but it doesn't seem to matter much in practice.
```
> synchronous XHR is not necessarily the best but it doesn't seem to matter much in practice.

I really really don't want to make assumptions like those of users. Not everyone has great internet, and we're using a third party that might not be reliable. I have great internet and a gaming PC, and I can notice this stutter.

irystocratanon replied:
I managed to get async fetch working with minimal changes

TuxedoTako commented on src/Linkification/Embedding.js:
```diff
@@ -592,30 +592,51 @@ var Embedding = {
+            <hr/>
+            &mdash; ${E(tweet.author.name)} (@${E(tweet.author.screen_name)}) ${created_at}
+            <br/>
+            🗨️${tweet?.replies || 0}&nbsp;🔄${tweet?.retweets || 0}&nbsp;❤️${tweet?.likes || 0}
```
I've had problems with the compatibility of unicode icons in the past: https://github.com/TuxedoTako/4chan-xt/issues/38.

irystocratanon replied:
Not sure what to do about this. Should I add new icons to src/Icons/icon.ts?

TuxedoTako replied:
You can use the comment and refresh icons already used in the header: ![msedge_s0nQ6vKpeg](https://github.com/TuxedoTako/4chan-xt/assets/126348645/8ab300c2-24f1-45c4-974e-e461f244f775), so only the hearth needs to be added.

TuxedoTako replied:
Another thing that I just noticed, `tweet?.replies`, `tweet?.retweets`, and `tweet?.likes` are not escaped.

irystocratanon replied:
They should be coerced to a number at all times (they already are but in the case of a malicious string we'd get `NaN || 0` instead of the string). I'll fix that.

TuxedoTako commented on src/Linkification/Embedding.js:
```diff
@@ -592,30 +592,51 @@ var Embedding = {
+            const mediaItem = mediaItems[i];
+            switch (mediaItem.type) {
+              case 'photo':
+                media += `<a target="_blank" href="${mediaItem.url}"><img src="${mediaItem.url}" style="max-width: 80vw; max-height: 80vh;"></a>`
```
Careful with injecting data from a third party API into the page. I see mediaItem.url isn't escaped.

irystocratanon replied:
Done. This was on purpose but I don't see any harm in escaping these.

TuxedoTako commented on src/Linkification/Embedding.js:
```diff
@@ -592,30 +592,51 @@ var Embedding = {
+
+          let {created_at} = tweet;
+          created_at = new Date(created_at);
+          created_at = created_at.toLocaleString(undefined, { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' });
```
```suggestion
          const created_at = new Date(tweet.created_at).toLocaleString(undefined, { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' });
```

While this file is still .js, I would prefer future code not to assign different types to the same variable. I know a lot of legacy code still does that, but this is new code.

TuxedoTako reviewed (changes requested):
It's nice to see a pull request instead of an issue, so I'm sorry to disappoint.

- I personally have not noticed problems with the twitframe:
  
  ![msedge_udjnWHLbW3](https://github.com/TuxedoTako/4chan-xt/assets/126348645/25183d78-dbc0-4bc2-a461-ea9a5a71780b)
  
  It might be because I don't usually browse boards with links to nsfw tweets. Maybe you can provide some examples that don't load?
- This implementation lacks some things: the one tweet I tested was a quote tweet, the current twitframe shows the quoted tweet, but this doesn't, despite the API returning the data.
-  Since a lot of twitter trouble is related to rate limits, do we have permission to direct traffic to `https://api.fxtwitter.com`? Does it have a licence?

irystocratanon committed:
Use async/fetch instead of blocking XHR

irystocratanon commented:
> It's nice to see a pull request instead of an issue, so I'm sorry to disappoint.
> 
>     * I personally have not noticed problems with the twitframe:
>       It might be because I don't usually browse boards with links to nsfw tweets. Maybe you can provide some examples that don't load?
It happens when signed out of Twitter a lot.
> 
>     * This implementation lacks some things: the one tweet I tested was a quote tweet, the current twitframe shows the quoted tweet, but this doesn't, despite the API returning the data.
> 
I should be able to fix that. I'll look into it. Another thing I don't think is handled right now are polls but that data should be in the API too I think.
>     * Since a lot of twitter trouble is related to rate limits, do we have permission to direct traffic to `https://api.fxtwitter.com`? Does it have a licence?
The license is here:
https://github.com/FixTweet/FxTwitter/blob/main/LICENSE.md

I think that is only for the code but they have docs here:
https://docs.fxtwitter.com/en/latest/api/about.html

They encourage use of their API.

irystocratanon committed:
Handle quotes

irystocratanon commented:
It handles quotes now. One thing I'm not sure about is nested quotes (a quote that quotes a quote that quotes a quote). I could handle that if the API has the data and I can find a tweet to test with.

irystocratanon committed:
Handle polls

TuxedoTako commented:
> It happens when signed out of Twitter a lot.

I don't even have a twitter account.

>I think that is only for the code but they have docs here:
>https://docs.fxtwitter.com/en/latest/api/about.html
>
>They encourage use of their API.

That does indeed look like permission.

TuxedoTako commented on src/Linkification/Embedding.js:
```diff
@@ -592,30 +596,92 @@ var Embedding = {
+            const created_at = renderDate(tweet);
+
+            const innerHTML = `
+              <p lang="${tweet.lang}" dir="ltr">${E(tweet.text)}</p>
```
Another unescaped variable. Maybe use the jsx function? That escapes automatically.

irystocratanon committed:
Add heart icon

irystocratanon committed:
Fixes

Escape more variables (I think I got them all this time)
Replace Emoji with SVG icons
Call Linkify to process links (In theory we can also run the embedder to
to embed thing like YouTube videos but that might be overkill and I
don't know how to do that)

irystocratanon committed:
Preserve linebreaks

irystocratanon committed:
Don't send referrer for img tags

Unfortunately this is still required for videos

irystocratanon committed:
Add configuration for fxTwitter

* Add a preference to control whether or not to translate Tweets
* I'm going to restore the old code in another commit to make this new
  code optional in case somebody wants to keep using twitframe

irystocratanon committed:
Translate tweets

irystocratanon committed:
Move config underneath Embedding

Images/Videos is not the right place for this

irystocratanon committed:
Restore old twitframe code and use this conditionally

irystocratanon committed:
Translate quotes

irystocratanon committed:
Resolve replies

irystocratanon committed:
Fix resolving replies

I iterated the wrong way so replies were displayed incorrectly

irystocratanon committed:
Check if we shouldResolveReplies

TuxedoTako commented:
I'm sensing scope creep. You're adding features after you already opened the pull request. I don't know where to test the reply resolver, which the original twitframe didn't have in the first place. But I already found an issue with the translation: if it's on a quote tweet, it's displayed as: quote tweet -> quoted tweet -> quoted translation -> quote translation.

irystocratanon commented:
>I don't know where to test the reply resolver

It's for threads. It didn't show any replies (the history, what it's in reply to) before and arguably doesn't need to (which is why I made it configurable) but this can be useful for context.

> it's displayed as: quote tweet -> quoted tweet -> quoted translation -> quote translation.

Do you mean the quote translation should be displayed first? That shouldn't be too difficult.

TuxedoTako commented:
> It's for threads. It didn't show any replies (the history, what it's in reply to) before and arguably doesn't need to (which is why I made it configurable) but this can be useful for context.

Do you have a thread where I can test this? I want to test before I merge the PR.

> > it's displayed as: quote tweet -> quoted tweet -> quoted translation -> quote translation.
> 
> Do you mean the quote translation should be displayed first? That shouldn't be too difficult.

I mean the translations should be grouped with the original text. Now the original tweet is sandwiched between the quote and its translation.

TuxedoTako commented:
I haven't found a tweet for 'Resolve all Tweet Replies', but I'm impatient, so I'm merging anyway.
I will move the templating to jsx, in case we missed an escape. I fixed the translation order myself.
